Add targets/stages argument to dvc exp show (#1)
iterative/enhancement-proposals: proposals/exp-show-target.md

- Enhancement Proposal PR: (leave this empty)
- Contributors: dberenbaum, (add your Github handle)

> # Summary

Reviewer: Meta: should there be a proposal title (in the template)?

Reviewer: Minor since obvi the PR has one but when you see the rendered MD doc it's not so evident.

Reviewer: Metameta: I can't resolve conversations in this repo @dberenbaum but feel free to collapse as you see fit.

The `dvc exp show` command shows a comparison of the results of different
experiments along with the parameters and metrics values for each experiment.

> The output currently includes all values that are defined in `params.yaml` or
> any other referenced parameters files in any DVC stage. Output also includes all

Reviewer: Sorry, prob no need for copy edits here but this may be important for disambiguation (metrics are DVC "outputs")
```suggestion
The command's output currently includes all values that are defined in `params.yaml` or
any other referenced parameters files in any DVC stage. It also includes all
```

> any other referenced parameters files in any DVC stage. Output also includes all

Reviewer: > or any other referenced parameters files in any DVC stage

~~I think only the specific params referenced in the stage are included? (not all the params in the referened params file)~~

It also includes values from `vars` in dvc.yaml I think.

Reviewer: > only the specific params referenced in the stage are included

Hmmm I guess not, per iterative/dvc/issues/5451.

Author: > It also includes values from `vars` in dvc.yaml I think.

Do you have an example of what you mean? Even if `vars` is used for substitution, wouldn't the substituted values still need to be parameters defined in `params.yaml` or another params file?

Reviewer: Ah yeah never mind, I was confused 👍

> any other referenced parameters files in any DVC stage. Output also includes all
> metrics in any referenced metrics files in any DVC stage. This proposal allows

Reviewer: Took me a bit to parse that sentence. Maybe this is slightly clearer:
```suggestion
any other referenced parameters files in any DVC stage. It also includes all
values in any referenced metrics files in any DVC stage. This proposal allows
```

Author: I broke this into bullets. Hopefully it's clearer.

users to see only parameters and metrics relevant to specified pipeline stages.

> See https://github.com/iterative/dvc/issues/5451 for more background.

Reviewer: What's the difference between this proposal and https://github.com/iterative/dvc/issues/5451? Other than the nice format and summary, of course!

Reviewer: Perhaps this was extracted from https://github.com/iterative/dvc/issues/5451#issuecomment-791205940 and that issue is focused on introducing the `--only-deps` flag?

Author: Yeah, the idea was to keep iterative/dvc#5451 focused on the short-term fix of adding an option to show only tracked parameters in `dvc metrics/params/exp show`. This proposal builds off the discussion there to introduce a potentially larger change specifically for `dvc exp show` (adding an argument for the stage of interest and only showing the parameters and metrics associated with that stage). In part, this is a trial run to see the pros and cons of moving larger feature changes into enhancement proposals, although this one is still pretty small by design.

Reviewer: > adding an option to show only tracked parameters

Right that's what confused me, because "This proposal allows users to see only parameters and metrics relevant to specified pipeline stages," which seemed quite similar to iterative/dvc/pull/5550.

> potentially larger change specifically for dvc exp show (adding an argument for the stage of interest...

OK I see so `exp show clean` would display all the same experiments as rows, but only metrics and params for the `clean` stage as rightmost columns?

Author: > OK I see so `exp show clean` would display all the same experiments as rows, but only metrics and params for the `clean` stage as rightmost columns?

Correct. An "unresolved question" at the bottom of the doc is whether to include only the relevant experiments as rows.

# Add targets/stages argument to dvc exp show

- Enhancement Proposal PR: (leave this empty)
- Contributors: dberenbaum, (add your Github handle)

# Summary

The `dvc exp show` command shows a comparison of the results of different
experiments along with the parameters and metrics values for each experiment.
The command's output currently includes:
* All values that are defined in `params.yaml`.
* All values in other referenced parameters files in any DVC stage.
* All values in any referenced metrics files in any DVC stage.
This proposal allows users to see only parameters and metrics relevant to

> specified pipeline stages.

Reviewer: > specified pipeline stages

But `targets` would be an optional command argument right (dvc.yaml by default) ? Meaning that the user could still use `exp show` without specifying anything.

Reviewer: Meta: also this needs a new line or will be interpreted as part of the last bullet:

```suggestion

This proposal allows users to see only parameters and metrics relevant to
specified pipeline stages.
```

See https://github.com/iterative/dvc/issues/5451 for more background.

# Motivation

`dvc exp show` is the main view by which users may compare experiments, and it's

> becoming useful for other purposes, like comparing multiple commits (see

Reviewer: > it's becoming useful for other purposes, like comparing multiple commits

Maybe we should discuss that too/separately (as a proposal) 🙂 Should this be a top-level command? Perhaps `dvc compare (--exp[eriments])`

Although, `dvc compare` is also the name I proposed to unify the different `diff` subcommands at the top-level (another possibly interesting proposal doc).

Reviewer: > discuss that too/separately (as a proposal)

Meta: how do we pick between / when to we move from https://github.com/iterative/dvc/discussions to https://github.com/iterative/enhancement-proposals ?

Author: > Meta: how do we pick between / when to we move from https://github.com/iterative/dvc/discussions to https://github.com/iterative/enhancement-proposals ?

Yes, I'm struggling with that already! Let's discuss at retro.

Reviewer 2: This really belongs in a separate discussion, but it seems like we will eventually want to unify all of the `diff` commands.

`exp diff` exists as a convenience wrapper for params + metrics diff, but all of the individual `dvc ... diff` commands (including `dvc diff` itself) already support comparing any git commit or reference (including shortened DVC experiment names). It would probably make more sense to just have one `diff` command and where you can specify which output (data/metrics/params) you want to include/exclude (and nothing about this would be specific to experiments).

Author: Funny you say that: https://github.com/iterative/enhancement-proposals/tree/unified-show-diff 😉

│ 72ed9cd        │ Nov 15, 2020 │  0.5674 │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 50                 │ -                   │ -
│ ├── exp-47dfa  │ Feb 17, 2021 │ 0.51625 │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 10                 │ -                   │ -
│ └── exp-44136  │ Feb 17, 2021 │  0.5674 │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 50                 │ -                   │ -
│ f8e9d93        │ Nov 15, 2020 │ 0.54175 │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 50                 │ -                   │ -
```

This table view is from a simplified example and is still missing several
columns, so this view can become cluttered quickly. Eliminating irrelevant
information is a priority.

There are several ways in which the included parameters might be irrelevant:
* Parameters files may include values that aren't tracked by DVC at all (for

>   example, debug level).

Reviewer: Same for metrics, I think (should they be mentioned?).

Reviewer 3: I think that in case of metrics we don't have this problem: https://github.com/iterative/dvc/issues/5451#issuecomment-796674685
because metrics are files while params are loaded as files while they should be loaded as file entries/lines.

Reviewer: So is the earlier bullet about metrics correct?

> The command's output currently includes: ...
> * All values in any referenced metrics files in any DVC stage.

Author: All metrics are outputs from a DVC stage, which means that:
* Unlike parameters, metrics are always tracked by some DVC stage.
* Metrics still may be irrelevant to a particular DVC stage, as mentioned in the bullets below.

                          *
                     +---------+
                     | prepare |
                     +---------+
                          *
                          *
                          *
                    +-----------+
                    | featurize |
                  **+-----------+**
              ****       *         ****
          ****           *             ***
        **              *                 ***
+-------+               *                    **
| train |             **                      *
+-------+            *                        *
         **        **                         *
           **    **                           *
             *  *                             *
        +----------+                  +----------------+
        | evaluate |                  | topic_modeling |
        +----------+                  +----------------+
                   ***             ****
                      *         ***
                       **     **
                     +---------+
                     | combine |
                     +---------+
```

There are no apparent reasons that users want to see untracked parameters in

> `dvc exp show`. Unlike `dvc params diff` and `dvc metrics diff`, `dvc exp show`

Reviewer: > There are no apparent reasons that users want to see untracked parameters in
`dvc exp show`.

This again brings me back to iterative/dvc/pull/5550. Probably that behavior should be the default, but seems like a separate question to adding `exp show [targets]`? Just trying to clearly separate concerns I guess.

Reviewer: see that PR is linked to close https://github.com/iterative/dvc/issues/5451. If that's the case then keeping the discussion about default behavior here makes sense, as it's closely related. (Pls resolve)

is not applicable to arbitrary files. Likewise, there are no apparent reasons
that users want to see parameters or metrics from irrelevant stages of their
pipelines. Users could keep untracked parameters in other files if they don't
want them shown in `dvc exp show`, and they could exclude parameters or metrics
from irrelevant stages, but DVC already has enough information to understand
that these should not be shown.

# Detailed design

`dvc exp show` could take stages as (positional?) arguments, similar to `dvc exp
run`. The help message could be similar:

```console
positional arguments:

> stages               Stages for which to print experiment info. 'dvc.yaml'
> by default.

Reviewer: I'd go with `targets` for consistency with existing commands (e.g. [`repro`](https://dvc.org/doc/command-reference/repro)), and since dvc.yaml files can be targets too.

```suggestion
usage: dvc experiments show [-h] [-q | -v] [-a] [-T] [-A] [-n <num>]
                            ...
                            [targets [<target> ...]]

targets              Stages to include. 'dvc.yaml' by default.
```

Author: What about `dvc params/metrics diff --targets`, which use `targets` to mean filenames and not stages?

Reviewer: I'm thinking of existing commands that target stages (again like `repro` which predate params and metrics BTW). But yeah `--targets` for some reason was designed to specifically accept file names and not param/metric stages, maybe something else to reconsider but seems unrelated here 🙂

```

If no stages are passed, parameters and metrics defined in any stage would be

> shown. This differs from current behavior only because untracked parameters

Reviewer: > defined in any stage

Any stage in ./dvc.yaml only or any stage anywhere (like `repro -P`?

Author: Good point. I think any stage anywhere is better, but we need to work on terminology for this.

Reviewer: I think the default targets for `repro` etc is just ./dvc.yaml

Author: Ah, thanks for pointing that out. In that case, I guess we should have a default of `./dvc.yaml` and include a `-P, --all-pipelines` option. Seems like maybe I need to go through the repro options and determine all that should apply here.

would be excluded.

If stages are passed, parameters and metrics defined anywhere in the pipeline

> for those stages would be shown.

Reviewer: Why not just for those specific stages? Wouldn't mind incorporating the `--[up/down]stream` options to this proposal TBH. Full makeover 💇 💅

Reviewer: Or `--single[-item]` as mentioned a bit further down.

Reviewer: > this is out of scope for now

The proposal format seems a bit long anyway, so why not cover as much as possible? Reuse the motivation

Author: > The proposal format seems a bit long anyway, so why not cover as much as possible? Reuse the motivation

True, I'm used to trying to limit scope for issues, but one advantage of this format is to holistically discuss related issues in one cohesive doc. Thoughts on `--single[-item]`?

Reviewer: > Thoughts on --single[-item]?

If the idea here is to remove noise from the very busy `exp show` table maybe single-item should be the default and have `--upstream` instead (or some other names). No strong opinion though.

Would also throw in `--downstream` for consistency in either case.

Author: Again, my initial reaction is to follow the `repro` conventions, which would do upstream by default, but also include:

* `-s, --single-item`
* `-p, --pipeline`
* `--downstream`

Reviewer: No strong opinion

│ c778a78        │ Nov 15, 2020 │       - │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 10                 │ -                  │ - │
│ 551082e        │ Nov 15, 2020 │       - │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 10                 │ -                  │ - │
│ 15bef96        │ Nov 15, 2020 │       - │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 10                 │ -                  │ - │
└────────────────┴──────────────┴─────────┴───────────────┴──────────────┴────────────────────────┴──────────────────┴────────────┴────────────────────┴────────────────────┴───┘
```

This table includes parameters and metrics from the `evaluate` stage and all
stages upstream of it.  It's possible to also have a `--single-item` argument to
exclude upstream changes, but this is out of scope for now.

# How We Teach This

See [Detailed design](#detailed-design) for a proposed help message.

This is consistent with other commands, and it's only one additional argument to

> an existing command, so it likely doesn't require much teaching. It might be

Reviewer: For this section I think something v useful would be to list the specific docs that could be affected by this (and how), and possibly new ones too.

Reviewer: BTW lmk if you'd like me to propose a list.

Author: 🙏 That would be great!

necessary to explicitly add that parameters and metrics from upstream stages
will be shown.

# Drawbacks

This proposal may require changes to how DVC collects and tracks parameters and
metrics. Implementation could break existing functionality, and it is
inconsistent with the current behaviors of `dvc params`, `dvc metrics`, and `dvc

> exp diff` commands.

Reviewer: This may be an advantage. Otherwise why have all these different options if they do the same thing?

Reviewer 3: > This may be an advantage.

I agree though we need to remember that we released a major version recently, and this would be kind of a breaking change. We don't know how many users actually want to see current behavior, and if changing it won't make users lose trust in our development process.

Reviewer 3: Ok I guess its addressed down below.

> Further, this proposal adds to an already crowded set of `dvc exp show` options,

Reviewer: Another related-discussion/proposal: can we simplify the flags? E.g. merge the `--include`, `--exclude`, and `--sort` ones (may require them to accept 2 CLI args which I'm not sure is possible). E.g. `exp show --inc :all auc,rank --sort p.tsv:epochs asc` would include all params and just those 2 metrics, while inverse sorting by the `epochs` param.

Reviewer: Or at least shorten the names. In any case, not sure it's so crowded in terms of functionality, since these 6 flags are all for similar purposes. Similar to [`dvc plots show`](https://dvc.org/doc/command-reference/plots/show) which also has a bunch of display-related options.

Reviewer: p.s. similar to [`dvc plots show`](https://dvc.org/doc/command-reference/plots/show) which also has a bunch of display-related options.

and there is no single way to show parameters and metrics that's going to work
for every use case. For example, this proposal does nothing to filter out
parameters for upstream stages, which may frequently be irrelevant if only the
stages near the end of the pipeline change during experimentation.

# Alternatives

There are many alternative methods to refine the output of `dvc exp show`:

* Current functionality: `exp show` already has options to include/exclude any
  column, so the flexibility to get the same output as suggested here already

>   exists.

Reviewer: But DVC still has to read all those stages and values whereas with `targets` the operation could actually improve performance-wise (perhaps).

* Expand the include/exclude options to accept wildcard/glob/regex arguments for

>   more flexibility to exclude an entire file or section of a file like

Reviewer: Def. Another good proposal to include 👍 (or make an issue in the core repo directly?)

Reviewer: > `exclude-params=params.json:local_config*`

This specific case may already be possible? (Without the wildcard) by specifying a group/parent param path.

Author: > Def. Another good proposal to include 👍 (or make an issue in the core repo directly?)

Added links to existing core dvc issues with some of these proposals.

> This specific case may already be possible? (Without the wildcard) by specifying a group/parent param path.

I don't think it works without a fully specified param name, but let me know if you find otherwise.

Reviewer: Seems to work:

```console
$ dvc exp show --no-pager
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━┓
┃ Experiment    ┃ Created      ┃ p.one ┃ p.two ┃ q ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━┩
│ workspace     │ -            │ 1     │ 2     │ q │
│ master        │ Mar 01, 2021 │ 1     │ 2     │ q │
│ └── exp-d662e │ Mar 01, 2021 │ 1     │ 2     │ q │
└───────────────┴──────────────┴───────┴───────┴───┘
$ dvc exp show --no-pager --include-params p
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┓
┃ Experiment    ┃ Created      ┃ p.one ┃ p.two ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━┩
│ workspace     │ -            │ 1     │ 2     │
│ master        │ Mar 01, 2021 │ 1     │ 2     │
│ └── exp-d662e │ Mar 01, 2021 │ 1     │ 2     │
└───────────────┴──────────────┴───────┴───────┘
$ dvc exp show --no-pager --include-params p.two
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━┓
┃ Experiment    ┃ Created      ┃ p.two ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━┩
│ workspace     │ -            │ 2     │
│ master        │ Mar 01, 2021 │ 2     │
│ └── exp-d662e │ Mar 01, 2021 │ 2     │
└───────────────┴──────────────┴───────┘
```

Author: Oh, nice! I had tried with just a filename, which doesn't work, but I guess a yaml section does work.

Reviewer 2: The current include/exclude filtering implementation is naive - filtering is done at the UI level based on string comparisons for the column names.

Filtering by filename will work for cases where the table actually displays the filenames (meaning whenever a duplicate key exists in multiple files). But in normal cases where we do not need to display filenames, using `filename:` won't do anything.

So if your repo has `params1.json:foo` and `params2.json:foo`, the table will include the filenames as part of the column title, and `exp show` will also respect the `filename:` part of a column filter.

Author: See also https://github.com/iterative/dvc/pull/5720

  `exclude-params=params.json:local_config*`.
* Save configuration options for `dvc exp show` so they can be reused and

>   modified without typing out all options at the command line each time.

Reviewer: Another great idea. BTW also similar to `plots show/diff` (plots templates).

Reviewer 3: Just yesterday we talked about being able to group the plots into "report files".

* Add a tsv/csv output format option in `dvc exp show` to enable users to

>   manipulate the table with other tools.

Reviewer: ➕

Reviewer 3: :heavy_plus_sign:

Reviewer: Although... Since there's already `--show-json`, TSC/CSV may be too much (you can easily [convert](https://json-csv.com/) from JSON to tabular).

Author: > Although... Since there's already `--show-json`, TSC/CSV may be too much (you can easily [convert](https://json-csv.com/) from JSON to tabular).

This isn't true for `dvc exp show`, since the json is not structured like tabular data. Try it out and the results are kind of a mess (too messy to post in a comment here). The json could be cleaned up to make this easy, but it was specifically designed as output for downstream apps to consume instead of as a format that users can easily understand. So I guess it's less about having a csv format specifically, and more about having an easily understood output format that can be saved and parsed.

This proposal does not preclude any of these options. The current output
will continue to be available except for untracked parameters. The other
alternatives may still be implemented.

Each of the alternatives provide flexibility to modify the `dvc exp show`
outputs, which is important but not the goal of this proposal. Instead, the

> proposed feature uses the information DVC has to make smart decisions about what

Reviewer: Sounds like this def. covers the discussion about default behavior then, which I think is good (but may need clarification earlier on in the doc).

to show and save the users from needing to make modifications that would always
be desirable.

# Unresolved questions

> * What about consistency with `parameters`, `metrics`, and `exp diff` commands?

Reviewer: Meta: maybe merge this section with **Drawbacks**?
